src/gui/instrument/EnvelopeGraph.cpp

```diff
@@ -54,102 +56,208 @@ void EnvelopeGraph::modelChanged()
+	};
+	
+	addScalingEntry(tr("Dynamic"), tr("Uses absolute spacings but switches to relative spacing if it's running out of space"), ScalingMode::Dynamic);
+	addScalingEntry(tr("Absolute"), tr("Provides enough potential space for each segment but does not scale"), ScalingMode::Absolute);
+	addScalingEntry(tr("Relative"), tr("Always uses all of the available space to display the envelope graph"), ScalingMode::Relative);
+
+    menu.exec(event->globalPos());
 }
 
 void EnvelopeGraph::paintEvent(QPaintEvent*)
 {
-	QPainter p(this);
+	QPainter p{this};
 	p.setRenderHint(QPainter::Antialiasing);
 
 	// Draw the graph background
 	p.drawPixmap(rect(), m_envGraph);
 
-	const auto * params = castModel<EnvelopeAndLfoParameters>();
-	if (!params)
-	{
-		return;
-	}
+	const auto* params = castModel<EnvelopeAndLfoParameters>();
+	if (!params) { return; }
 
+	// For the calculation of the percentages we will for now make use of the knowledge
+	// that the range goes from 0 to a positive max value, i.e. that it is in [0, max].
 	const float amount = params->getAmountModel().value();
+
 	const float predelay = params->getPredelayModel().value();
+	const float predelayPercentage = predelay / params->getPredelayModel().maxValue();
+
 	const float attack = params->getAttackModel().value();
+	const float attackPercentage = attack / params->getAttackModel().maxValue();
+
 	const float hold = params->getHoldModel().value();
+	const float holdPercentage = hold / params->getHoldModel().maxValue();
+
 	const float decay = params->getDecayModel().value();
+	const float decayPercentage = decay / params->getDecayModel().maxValue();
+
 	const float sustain = params->getSustainModel().value();
+
 	const float release = params->getReleaseModel().value();
+	const float releasePercentage = release / params->getReleaseModel().maxValue();
+
+	// The margin to the left and right so that we do not clip too much of the lines and markers
+	const float margin = 2.0;
+	const float availableWidth = width() - margin * 2;
 
-	const float gray_amount = 1.0f - fabsf(amount);
+	// Now determine the maximum width for one segment according to the scaling setting.
+	// The different scalings use different means to compute the maximum available width per segment.
+	const auto computeMaximumSegmentWidthAbsolute = [&]() -> float
+	{
+		return availableWidth / 5.;
+	};
 
-	p.setPen(QPen(QColor(static_cast<int>(96 * gray_amount),
-				static_cast<int>(255 - 159 * gray_amount),
-				static_cast<int>(128 - 32 * gray_amount)),
-									2));
+	const auto computeMaximumSegmentWidthRelative = [&]() -> float
+	{
+		const float sumOfSegments = predelayPercentage + attackPercentage + holdPercentage + decayPercentage + releasePercentage;
 
-	const QColor end_points_color(0x99, 0xAF, 0xFF);
-	const QColor end_points_bg_color(0, 0, 2);
+		if (sumOfSegments != 0.)
+		{
+			return availableWidth / sumOfSegments;
+		}
+		else
+		{
+			return computeMaximumSegmentWidthAbsolute();
+		}
```

Comment:
Maybe?
```suggestion
		return sumOfSegments != 0.
			? availableWidth / sumOfSegments
			: computeMaximumSegmentWidthAbsolute();
```

Reply:
I have also kept this as is because the current version somehow lends more importance to the check for 0 before we divide even if logically the same is done in your proposal.

Reply:
The importance is the same. It's just less line usage.

Reply:
Ok, done with commit 26d260e7650.